packages/daimo-api/src/contract/foreignCoinIndexer.ts

```diff
@@ -0,0 +1,209 @@
+  }
+
+  processSwapCoinDelta(
+    addr: Address,
+    delta: bigint,
+    log: ForeignTokenTransfer
+  ) {
+    if (delta < 0n) {
+      // outbound transfer
+      this.sendsByAddrTxHash.set(addrTxHashKey(addr, log.transactionHash), log);
+
+      const pendingSwaps = this.pendingSwapsByAddr.get(addr) || [];
+
+      // Delete the first matching pending swap that is now swapped
+      const matchingPendingSwap = pendingSwaps.find((t) => t.value === -delta)!;
+
+      this.correspondingReceiveOfSend.set(
+        addrTxHashKey(addr, log.transactionHash),
+        matchingPendingSwap
+      );
+
+      const newPendingSwaps = pendingSwaps.filter(
+        (t) => t.transactionHash !== matchingPendingSwap.transactionHash
+      );
+
+      this.pendingSwapsByAddr.set(addr, newPendingSwaps);
+    } else {
+      // inbound transfer, add as a pending swap
+      if (this.pendingSwapsByAddr.has(addr)) {
+        this.pendingSwapsByAddr.get(addr)!.push(log);
+      } else {
+        this.pendingSwapsByAddr.set(addr, [log]);
+      }
+    }
+  }
+
+  /** Listener invoked for all past ETH transfers, then for new ones. */
+  pipeAllTransfers(listener: (logs: ForeignTokenTransfer[]) => void) {
+    listener(this.allTransfers);
+    this.addListener(listener);
+  }
+
+  /** Listener is invoked for all new ETH transfers. */
+  addListener(listener: (logs: ForeignTokenTransfer[]) => void) {
+    this.listeners.push(listener);
+  }
+
+  /** Unsubscribe from new ETH transfers. */
+  removeListener(listener: (logs: ForeignTokenTransfer[]) => void) {
+    this.listeners = this.listeners.filter((l) => l !== listener);
+  }
+
+  async getProposedSwapForLog(
+    log: ForeignTokenTransfer
+  ): Promise<ProposedSwap | null> {
+    const swap = await retryBackoff(`getProposedSwapForLog`, async () => {
+      const fromAcc = await this.nameReg.getEAccount(log.from);
+      return this.uc.getProposedSwap(
+        log.to,
+        log.value.toString() as `${bigint}`,
+        log.foreignToken,
+        guessTimestampFromNum(log.blockNumber, chainConfig.daimoChain),
+        fromAcc
+      );
+    });
+
+    console.log(
+      `[SWAPCOIN] getProposedSwapForLog ${log.from}: ${JSON.stringify(swap)}`
+    );
+    return swap;
+  }
+
+  async getProposedSwapsForAddr(addr: Address): Promise<ProposedSwap[]> {
+    const pendingSwaps = this.pendingSwapsByAddr.get(addr) || [];
+    const swaps = (
+      await Promise.all(
+        pendingSwaps.map((swap) => this.getProposedSwapForLog(swap))
+      )
+    ).filter((s): s is ProposedSwap => s != null);
+
+    return swaps;
+  }
+
+  // Used to detect onchain swaps initiated by the user.
+  getForeignTokenReceiveForSwap(
+    addr: Address,
+    txHash: Hex
+  ): ForeignTokenTransfer | null {
+    const log = this.sendsByAddrTxHash.get(addrTxHashKey(addr, txHash));
+    if (log == null) return null;
+
+    const correspondingReceive = this.correspondingReceiveOfSend.get(
+      addrTxHashKey(addr, txHash)
```

Comment:
👀

so this matches the in & out transfers of a swap based on (addr, txhash)?

should work for now, but won't in the general case, for example if we have a bigger bundle that contains two userops from the same Daimo account

open challenge, i guess, for Shovel to get us a more elegant way to do this.

shovel indexing internal contract calls would go a long way (eth transfers,  uniswap swaps, etc)

Reply:
> two userops from the same Daimo account

Yeah this seemed like enough of a future edge case that it seems fine to ignore. Will leave a comment.